### ISSUE TITLE
Feature flags - add support for variants

### ISSUE THREAD

**Issue by kpthatsme** [PostHog/posthog#5440]
## Is your feature request related to a problem?

Often when an experiment is rolled out, there are a few variants - at least two: the experiment and control.

Right now PostHog doesn't seem to have a good way of supporting multiple variants for a single flag. 

For example, let's say I want to send 50% of traffic into the new email  sign up experiment. We want 50% of traffic to be absolutely unchanged, and not be bucketed at all.

For the 50% in the experiment, we want to roll out Variant A, Variant B, and Control at 33% each. 

Currently, we don't have a great way of doing this in PostHog. We would need to count each individual variant as its own experiment and somehow tie all the context together so we bucket correctly. 

* note: You may wonder why we are allocating control in this 50% bucket as opposed to just saying anyone not in the experiment === control. That's fair, the reason for that is to support the notion of A/A, A/A/B testing– it's seen more of a push over the last few years to drive higher confidence results. [more](https://blog.analytics-toolkit.com/2014/aa-aab-aabb-tests-cro/) [reading](https://vwo.com/blog/aa-test-before-ab-testing/).

## Describe the solution you'd like

1. Create a feature flag like normal, allocating a % of traffic allocated to it and a set of rules that define what types of users can be bucketed into it.
2. Optionally, specify variants: define a variant name and a % of the feature flag's traffic that will be routed to this variant - note: this will always be <= the total allocation for the flag
3.  In our decide API, we will need to send back feature flag keys with an optional variant name if the flag has one set.

## Describe alternatives you've considered

Handle variants with a unique flag/ variant. 

I don't think there's a correct way to set this up in the product today and insure we're not double bucketing.

## Additional context

This definitely isn't a new problem or one specific to us - most feature flag solutions solve this so it's worth looking at how a few others handle this. My experience is mainly with LaunchDarkly.

**Comment by joesaunderson** [comment]
We would love this feature, I have been tracking https://github.com/PostHog/posthog/issues/3431 as I think it's similar, but I love the notion of a 'control' group from this issue.

**Comment by samwinslow** [comment]
[Here](https://github.com/PostHog/product-internal/issues/124#issuecomment-892356376) @kpthatsme mentioned a method to randomly assign people to groups when `identify` is called. I think we actually want to do it at the `decide` stage, right, so that it affects users who may not yet be identified (i.e. first-time site visitors).

**Comment by kpthatsme** [comment]
@samwinslow My proposal there was to unblock the current experiment we're trying to ship without needing built-in support for multi-variants tied to a single flag. 

I was not necessarily suggesting that should be the long-term solution for supporting this variant feature – it could be a piece of it, but I don't have enough context for how we've implemented feature flags to know if that's the right thing to do.

In regard to the experiment - I think our goal should be to get that shipped first and unblock ourselves through that proposal while we think about the priority of this. I'll have a chat with Marcus about this too in my 1:1.

**Comment by kpthatsme** [comment]
@samwinslow assigned this to you - thinking we could use this ticket to track the actual work done as it's public facing.

**Comment by samwinslow** [comment]
Some notes here about what needs to be done for multivariate support in `posthog-js`. Roughly sorted into high and low priority.

High Priority
- [x] isFeatureEnabled
- [ ] Register/unregister persistence for flag values
- [x] React Hooks

Low Priority
- [ ] `$override_feature_flags` and `override()`: typing and docs update

Administrative
- [ ] Release new `posthog-js` version after the next `posthog` version, as feature flags v2 will be a dependency
- [ ] Draft docs updates for all the above

**Comment by samwinslow** [comment]
Some considerations @mariusandra and I talked about today:

- `active_feature_flags` (event and user property) is currently an array of strings, and we want to keep it that way
  - otherwise users would have to edit their existing PostHog queries that may use `active_feature_flags` for filtering
  - Moving to an object would have adverse effects for users that both do and do not use multivariate flags.
  - Moving to an object may result in worse query performance as nested JSON will need to be parsed every time
- `active_feature_flags` should include the key name of a multivariate flag
  - Two logically identical definitions of "active" as I see it:
    - A flag is "active" if it is not false and not undefined, i.e. it is set. Different than "a flag is active if it is strictly true".
    - A flag is "active" if the user matches a rollout group.
  - it may be desirable to do less granular analysis than checking for exact variants.
  - Say, for instance, you have a variant flag rolled out to 50% of users, and there are 3 variants. If you want to compare the 50% of users that have **some value** against the 50% that have **no value**, it's most intuitive to use the active feature flags.

 > send each enabled feature flag as a separate property (automatically flattened)

- This means that, like with our signup experiment, where we had a property `experiment_variant: string`, a user would have a property for each flag, called something like `feature_flag__FLAG_NAME: string | boolean`.
  - This would contain the active variant key, or a true/false value.

The idea is to make multivariate flags and standard (Boolean) flags as interoperable as possible, especially since it is possible to change a flag's type from standard (Boolean) to multivariate and vice-versa. This design would minimize breaking changes and introduce more flexibility into querying.

@kpthatsme please let me know your thoughts!

**Comment by samwinslow** [comment]
Closing this out since the initial scope is released in alpha + this issue will get too long otherwise.
